### ISSUE TITLE
Thanks, and drawing triads in Director

### ISSUE THREAD

**Issue by dehann** [rdeits/DrakeVisualizer.jl#21]
Hi Guys,

Thanks for putting these libraries out there, this is pretty awesome (@rdeits, @tkoolen, @patmarion)! I'm going to actually start using this more, and will be integrating into some of our packages also -- emphasis on state estimate/SLAM.

I was wondering about using Director with classic axis triads we have in Collections Viewer? I tried publishing similar LCM messages to Collections Viewer visualization traffic, but did not get any triads showing up. I see mention in [Director documentation](https://openhumanoids.github.io/director/tutorials/visualization/frames_and_transforms.html), but not enough to get it drawing.

Backward compatibility is not as important, rather efficiency and ease of use (large numbers of triads, etc.). Point Clouds would be a next step. Would you suggest I wrap a primitive with three long narrow boxes, as a triad, and do a pull request back to you?

Regarding Point Clouds there is also the older conversation about ```Collections Render``` from Hordur. The idea is that a point cloud can be slaved to a pose. Updating the pose via LCM would then automatically update the point cloud on the visualization side, thereby minimizing LCM traffic. This becomes important pretty quickly when drawing a large map.

Here is what I have been doing thus far with Collections Viewer and draing from Julia (using ```PyLCM.jl```), using yet another transforms library ([TransformUtils.jl](https://github.com/dehann/TransformUtils.jl)):
```julia
function pubRealTimeEst(pidol, lstPosel::Pose3D)
  dummyxj = deepcopy(lstPosel)
  dummyxj.rnTime = round(Int64,pidol.iDtj*1e9)
  # latestPose3
  DXhat =residual(pidol, lstPosel, dummyxj) #2/3
  dxij = newPose3D()
  dxij.rRp = convert(SO3, so3(DXhat[1:3][:])).R
  dxij.rVelp = DXhat[7:9][:]
  dxij.rPosp = DXhat[10:12][:]
  Xnow = oplus(lstPosel, dxij)
  newobj = vs.obj_t()
  newobj[:id] = 1;
  newobj[:x] = Xnow.rPosp[1]
  newobj[:y] = Xnow.rPosp[2]
  newobj[:z] = Xnow.rPosp[3]
  E = convert(Euler, SO3(Xnow.rRp))
  newobj[:roll] = E.R
  newobj[:pitch] = E.P
  newobj[:yaw] = E.Y

  objs = [newobj];

  newobjcoll = vs.obj_collection_t();
  newobjcoll[:id] = 10;
  newobjcoll[:reset] = true;
  newobjcoll[:name] = "state-est";
  newobjcoll[:type] = 5;
  newobjcoll[:nobjs] = 1;
  newobjcoll[:objs] = objs;
  #
  publish(pido.lc, "OBJ_COLLECTION", newobjcoll)
  nothing
end
```

**Comment by dehann** [comment]
Maybe something [like this](https://github.com/dehann/Caesar.jl/blob/master/src/VisualizationUtils.jl), showing three rolling triads about x-axis?

**Comment by rdeits** [comment]
@dehann that's great! I'm glad this will be getting some use. 

I think drawing the triad as a set of boxes is OK, but it might be better to add TRIAD as another supported geometry class in the visualizer itself. That would entail:

1. adding another enum value to [the lcm type](https://github.com/openhumanoids/bot_core_lcmtypes/blob/master/lcmtypes/bot_core_viewer_geometry_data_t.lcm#L13) (this doesn't change the message signature, so we don't break compatibility)
1. handling that value in [drakevisualizer.py](https://github.com/RobotLocomotion/director/blob/master/src/python/director/drakevisualizer.py#L70). We can decide to, for example, use the float_data field in the message to store the x, y, and z lengths of the triad

That would have the advantage of exposing the API to all director users, even outside of Julia. In that case, all we'd have to do is add a `Triad` type in `DrakeVisualizer.jl` and create its `to_lcm()` method. 

As for the point clouds, I recently added point cloud support to `drakevisualizer.py` here: https://github.com/RobotLocomotion/director/blob/master/src/python/director/drakevisualizer.py#L310 but currently it doesn't support publishing a pointcloud and then changing its transform later. I think that we can easily add that feature, though. As I'm the only user of the current API, I'm happy to break it. How about two channels:

* `DRAKE_VIEWER_LOAD_POINTCLOUD` which sends a new message type with:
  * a string ID for the pointcloud
  * the pointcloud as a bot_core_pointcloud_t
* `DRAKE_VIEWER_DRAW_POINTCLOUD` which sends a new message with:
  * the ID of the pointcloud
  * the transform

Does that work for you? And @patmarion does that seem reasonable?

**Comment by patmarion** [comment]
sure.  also, I forget why there is a separate method for drawing pointclouds than link meshes?

**Comment by rdeits** [comment]
Agreed, it makes perfect sense to have pointclouds and meshes use the same code path. I didn't implement it that way because we're using pointcloud_t messages to encode the pointclouds, so we can't store them inside viewer_geometry_data_t messages. That meant creating a separate channel and path for handling pointclouds.

**Comment by rdeits** [comment]
In order to support loading pointclouds and then setting their transforms later, I'm going to need to break the existing pointcloud interface (OK, since I'm the only user) and create a new LCM type (annoying but not a problem). 

However, if we actually want to be able to treat pointclouds like any other link geometry, then we probably need to change the `viewer_link_data_t` message, which will break compatibility with the current protocol. I'm fine with that, but it's kind of a big change for Drake. However, it might also be a nice opportunity to build a better `drake-visualizer` protocol. 

A few things on my `drake-visualizer` wishlist:

* support for pointclouds on links
* making it easy to add additional geometry types that I haven't thought of yet in the future
* meaningful responses from the viewer for each command
  * e.g.: client sends DRAW_ROBOT, viewer responds with "I don't have that link", client can send LOAD_ROBOT to load the required links
  * we could even get fancy: e.g. server could identify each link by a hash of its viewer_link_data message, and the client could send drawing commands that reference those hashes and their poses. That way, we could eliminate the stateful robot loading entirely: the client just always sends draw commands, server responds with any hashes it's missing, client sends link data for those hashes. If the client or server dies it's no problem because the process just transparently starts over.

**Comment by rdeits** [comment]
Ok, actually I'm pretty excited about the link data caching thing. Forcing the client to be aware of the visualizer's state has always bothered me, and we could fix that with, I think, almost no loss in performance.

**Comment by dehann** [comment]
Okay great, it may be worth it to just have a brief look at [Collection Render](https://github.com/openhumanoids/pronto/blob/573421916e887038f78e0914181ee2459e5c380d/visualization/src/collections_renderer/collections_renderer.cpp#L914). Hordur from our group did precisely this for the DGC viewer a few years back and I think this is the code for it. 

***Regarding triads:***
I can start by expanding the LCM type as you suggested above? I am busy building out more functionality for myself, so that I can  draw meshes and Geometries together. I'm working with separate HyperRectangle for now.

First you define a reference pose. Then you define an update-able pose with unique ID and relative transform to reference ID: Something like ```:origin``` and pose ```:x1```. Anything defined relative to a pose is then updated as close to the viewer side as possible. The same plumbing can then be used to do point clouds as a cached update-able mesh, geometry or link.

***Regarding point clouds:***
My suggestion is to have the ability to:
* upload a pose with unique ID.
* upload link, mesh, geometry or point cloud against that pose ID, but also with its own data ID and relative transform. Imagine trying to upload point clouds from two cameras on the same pose as example. ```Dict{T,Link}``` or something like that.
* upload a new pose to the same pose ID and thereby have the pose and entire constellation of cached meshes, geometries, links attached to that pose move as well.
* some way of deleting that pose, thereby stripping out the cached info from the viewer also. 


Collections Render indexing can get a bit confusing at first, because it supports each of these cases.

Robin, to understand correctly: Do you suggest a change in Drake Visualizer side, or actually in Director? How can I help? Maybe get the code working Drake visualizer side and then migrate across the LCM channel?

**Comment by patmarion** [comment]
Director has support for the collections renderer, it's just not enabled in the build you guys are using.

**Comment by rdeits** [comment]
@dehann OK, I think I understand what you need w.r.t. point clouds. I like the reference pose idea, and in fact, I think it's totally compatible with the basic ideas of the drake-visualizer. Essentially, a `Link` in the visualizer corresponds exactly to a reference pose, since it can contain any number of geometries which have their own transforms w.r.t. the link, but can all be moved together by updating the Link's pose. 

The changes I'm suggesting about would have to be in both DrakeVisualizer.jl and in Director, and will involve a pretty big change to the `drake-visualizer` LCM API. But I think that they'll let you publish linked geometries/meshes/point clouds and update their poses in the way you're talking about. I've got about 50% of a prototype of the new interface, and I think I can finish the rest of the prototype tonight. Maybe tonight I can send you a more concrete description of what I'm proposing and you can see if it fits your use cases? 

And meanwhile, I won't be working on the triad rendering, so if you want to expand the LCM type to include a triad enum and handle that of the Director side in `drakevisualizer.py`, that would be super helpful. 

@patmarion is the collections renderer support vestigial or is it something we actually want to keep going forward? That is, would it be worth trying to use it to draw triads and such, or is it better to just keep adding various polyData forms the way `drakevisualizer.py` already does?

**Comment by dehann** [comment]
Here is a snippet of what I am working on, showing isocontours of constant probability density, similar to
![equation](http://www.sciweavers.org/tex2img.php?eq=1,2,3\sigma&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=) ellipse for pure Gaussian. Drawing from [same place as before](https://github.com/dehann/Caesar.jl/blob/master/src/VisualizationUtils.jl):
![](https://raw.githubusercontent.com/dehann/Caesar.jl/23f9387acaab870cf74fc98079571fe0e9dc0c54/doc/imgs/directorexample01.png)

**Comment by dehann** [comment]
Great thanks! I'll do the triads part in LCM then. I can help test and debug if you have something to try. I'll see if I can stop in by you guys tomorrow, if you are around.

**Comment by patmarion** [comment]
> @patmarion is the collections renderer support vestigial or is it something we actually want to keep going forward? That is, would it be worth trying to use it to draw triads and such, or is it better to just keep adding various polyData forms the way drakevisualizer.py already does?

I don't personally use collections and I didn't add the support myself, so I don't know much about it, or how complete the support is.  I know it is using a rendering style that does not play nice with the visualizer (similar to lcm gl) so I wouldn't recommend it as a renderer going forward.  But for folks with existing tools that publish collections messages, it might be relevant.

**Comment by rdeits** [comment]
Ok, here is a more concrete proposal:

I suggest creating the following LCM types:

```c
struct viewer2_load_links_t {
	int64_t timestamp;

	int32_t num_links;
	viewer2_link_data_t link_data[num_links];
}

struct viewer2_link_data_t {
	string path;

	int32_t num_geometries;
	viewer2_geometry_t geometries[num_geometries];
}

struct viewer2_geometry_t {
	string name;

	position_3d_t pose;

	???? data;
}

struct viewer2_draw_t {
	int64_t timestamp;

	int32_t num_links
	string paths[num_links];
	position_3d_t poses[num_links];
}

struct viewer2_path_list_t {
	int64_t timestamp;

	int32_t num_paths;
	string paths[num_paths];
}

struct viewer2_response_t {
	int64_t timestamp;

	int32_t status;
	const int32_t STATUS_OK;
	const int32_t STATUS_NO_SUCH_LINK;
	const int32_t STATUS_LINK_LOAD_ERROR;
	const int32_t STATUS_DELETE_ERROR;

	string json;
```

and introducing the following channels:

### Loading
```
VIEWER2_LOAD_LINKS: viewer2_load_links_t
```
For each link_data in the message, spawns the included geometries. The `path` field generalizes our current concept of `robot_id` and `link_name` by using a `/`-separate string with arbitrary depth. Each `/` in the path creates a folder in the object model, with the geometries belonging to that link being placed inside the folder specified by the path. This does not delete anything from the object model, but will overwrite any geometries which have exactly the same path and name. 

A single link_data can contain arbitrarily many geometries, all of which will be transformed together when the user sends a `DRAW` message. 

Returns: viewer2_response_t with `STATUS_OK` or `STATUS_LINK_LOAD_ERROR`

### Drawing
```
VIEWER2_DRAW: viewer2_draw_t
```
For each path, set the pose of link at that path to the given pose. I suggest that we only allow setting transforms for leaf paths. That is, if you create a link at `robot1/group2/link3`, then it's OK to set a pose for `robot1/group2/link3`, but not OK to set a pose for `robot1` or `robot1/group2`. We can decide to allow this later on if we want. 

Returns: viewer2_response_t with `STATUS_OK` or `STATUS_NO_SUCH_LINK` if at least one link is not present in the visualizer. If any links are missing, also includes their paths in the `json` field of the response. 

If a client receives `STATUS_NO_SUCH_LINK`, it should send an appropriate `LOAD_LINK` command. 

### Deleting
```
VIEWER2_DELETE: viewer2_path_list_t
```
For each path in the message, recursively delete that folder and all subfolders from the object model. 

Returns: viewer2_response_t with `STATUS_OK` or `STATUS_DELETE_ERROR`


### Usage

Since `LOAD` never deletes anything, it's easy to create a geometry and then add new geometries (on that link or on different links) later. But we can also recreate the current `LOAD_ROBOT` behavior (which clears all robots) by sending a `DELETE` and then a `LOAD`. 

The big question that I haven't answered is what exactly is the type that stores the actual geometry data (labeled as `????` above)? Do we try to create one LCM message with lots of optional fields? Or an LCM message with enough fields for our current use and a raw bytearray for future expansion? Or just encode all geometry information using something like msgpack or protobuf that will give us backwards compatibility when we add new geometry types in the future? @patmarion any thoughts? 

@dehann does this fit all of your use cases (assuming that we can figure out how to make pointclouds one of the possible geometry types)?

**Comment by rdeits** [comment]
I've pushed some work in progress to https://github.com/rdeits/director/tree/caching including a demo application that uses the new interface. The relevant magic line is here: https://github.com/rdeits/director/blob/caching/newviewer.py#L45 With that line, you can kill the visualizer at any time, and as soon as you open a new one the geometry is immediately reloaded. 

I haven't handled deleting geometries yet, and I haven't handled the problem of how to store the geometry types (currently I'm just not supporting pointclouds at all).

**Comment by patmarion** [comment]
Maybe a bytearray to hold data, and a string that says the name of the protocol to use to interpret the data.  Perhaps one more field to hold a version number (although that could be part of the protocol name).  Perhaps an array of bytearray rather than just one.  I'm sure you'll get more ideas once you attempt to support a few use cases.

**Comment by rdeits** [comment]
Sounds good. I'll play around with something like that.

**Comment by rdeits** [comment]
The more I work on this, the less convinced I am that LCM types are helping at all here. Is the logical extension just to convert the entire protocol to raw data storing JSON? We can keep the format name and version in the message so that later we can move to something like MsgPack if we need better performance.

**Comment by rdeits** [comment]
Ok, JSON interface is working, and it's awesome. Primitives, pointclouds, planar LIDARs, and meshes are all in (although I haven't actually tested meshes yet). The robot_id and link_name have been replaced with general paths for arbitrary organization. Pointclouds and LIDARs now support colorization by intensity or RGB and can be attached to links just like any other geometry (so they can be rigidly attached to other geometries in the same link if you want). It also has all the reloading magic, so you can kill the viewer and restart whenever you want. 

All that, and it's only a net gain of 16 lines of code 😉 

@patmarion I think this is going to be cool. I'll show you in lab tomorrow. 

https://github.com/rdeits/director/tree/json-interface

**Comment by dehann** [comment]
Hi @rdeits, wow this looks cool! Let me get back to you, its been a bit of whirlwind over Christmas... Happy new year!

**Comment by rdeits** [comment]
@dehann I've pushed a new branch here: https://github.com/rdeits/DrakeVisualizer.jl/tree/treeviewer that uses the new interface and supports all of the geometry types (plus, now, triads).

**Comment by dehann** [comment]
Hi @rdeits , I was finally able to get back to this :-P I got the triads drawing and converted my code to use the new tree visualizer interface -- it works well thanks! You might like to know I showed a few [visualizations](https://vimeo.com/198237738) at a conference recently.

Will hopefully get to using the point clouds soon.

**Comment by rdeits** [comment]
Great! I'm glad it's working!

**Comment by rdeits** [comment]
Closing this, since these features are all implemented in v0.1.0.
